src/wallet/wallet.cpp

```diff
@@ -3146,19 +3146,18 @@ bool CWallet::GetCollateralTxDSIn(CTxDSIn& txdsinRet, CAmount& nValueRet) const
 
     std::vector<COutput> vCoins;
 
-    AvailableCoins(vCoins);
+    AvailableCoins(vCoins, true, NULL, false, ONLY_PRIVATESEND_COLLATERAL);
 
-    for (const auto& out : vCoins)
-    {
-        if(CPrivateSend::IsCollateralAmount(out.tx->tx->vout[out.i].nValue))
-        {
-            txdsinRet = CTxDSIn(CTxIn(out.tx->tx->GetHash(), out.i), out.tx->tx->vout[out.i].scriptPubKey);
-            nValueRet = out.tx->tx->vout[out.i].nValue;
-            return true;
-        }
+    if (vCoins.empty()) {
+        return false;
     }
 
-    return false;
+    std::random_shuffle(vCoins.rbegin(), vCoins.rend(), GetRandInt);
```

Comment:
No need to use reversed iterators here, as selection is random nevertheless. Also, it implies some hidden logic which is not there actually.

Maybe we shouldn't use`random_shuffle` at all in such cases? In the end, we just want a random index of the vector to be chosen, so something like the following suggestion should be better.